src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -614,23 +618,105 @@ public void WithProjectDefaultNamespace()
             var projectId = ProjectId.CreateNewId();
 
             using var workspace = CreateWorkspace();
-            var solution = workspace.CurrentSolution
-                            .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+            var solution = workspace.CurrentSolution.
+                AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
 
             // any character is allowed
             var defaultNamespace = "\0<>a/b/*";
 
             SolutionTestHelpers.TestProperty(
                 solution,
                 (s, value) => s.WithProjectDefaultNamespace(projectId, value),
-                s => s.GetProject(projectId)!.DefaultNamespace,
+                s => s.GetRequiredProject(projectId).DefaultNamespace,
                 (string?)defaultNamespace,
                 defaultThrows: false);
 
             Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectDefaultNamespace(null!, "x"));
             Assert.Throws<InvalidOperationException>(() => solution.WithProjectDefaultNamespace(ProjectId.CreateNewId(), "x"));
         }
 
+        [Fact]
+        public void WithProjectChecksumAlgorithm()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            using var workspace = CreateWorkspace();
+            var solution = workspace.CurrentSolution.
+                AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            SolutionTestHelpers.TestProperty(
+                solution,
+                (s, value) => s.WithProjectChecksumAlgorithm(projectId, value),
+                s => s.GetRequiredProject(projectId).State.ChecksumAlgorithm,
+                SourceHashAlgorithms.Default,
+                defaultThrows: false);
+        }
+
+        [Fact]
+        public void WithProjectChecksumAlgorithm_DocumentUpdates()
+        {
+            var projectId = ProjectId.CreateNewId();
+            var documentAId = DocumentId.CreateNewId(projectId);
+            var documentBId = DocumentId.CreateNewId(projectId);
+            var documentCId = DocumentId.CreateNewId(projectId);
+            var documentDId = DocumentId.CreateNewId(projectId);
```

Comment:
Can we rename the Ids (or at least D) to something better describing what makes each unique?